File: src/test/java/no/nav/veilarboppfolging/service/AktiverBrukerIntegrationTest.java
package no.nav.veilarboppfolging.service;

import no.nav.common.types.identer.AktorId;
import no.nav.common.types.identer.Fnr;
import no.nav.veilarboppfolging.IntegrationTest;
import no.nav.veilarboppfolging.LocalDatabaseSingleton;
import no.nav.veilarboppfolging.client.amtdeltaker.AmtDeltakerClient;
import no.nav.veilarboppfolging.client.digdir_krr.KRRData;
import no.nav.veilarboppfolging.domain.Oppfolging;
import no.nav.veilarboppfolging.eventsLogger.BigQueryClient;
import no.nav.veilarboppfolging.oppfolgingsbruker.inngang.AktiverBrukerManueltService;
import no.nav.veilarboppfolging.repository.*;
import no.nav.veilarboppfolging.test.DbTestUtils;
import org.junit.Before;
import org.junit.Test;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.transaction.support.TransactionTemplate;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class AktiverBrukerIntegrationTest extends IntegrationTest {

    private OppfolgingsStatusRepository oppfolgingsStatusRepository;
    private AktiverBrukerManueltService aktiverBrukerManueltService;
    private ManuellStatusService manuellStatusService;
    private final Fnr FNR = Fnr.of("1111");
    private final AktorId AKTOR_ID = AktorId.of("1234523423");
    private BigQueryClient bigQueryClient;
    private KafkaProducerService kafkaProducerService;

    @Before
    public void setup() {
        JdbcTemplate jdbcTemplate = LocalDatabaseSingleton.INSTANCE.getJdbcTemplate();
        NamedParameterJdbcTemplate namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(jdbcTemplate);
        TransactionTemplate transactor = DbTestUtils.createTransactor(jdbcTemplate);
        oppfolgingsStatusRepository = new OppfolgingsStatusRepository(new NamedParameterJdbcTemplate(jdbcTemplate));
        oppfolgingsPeriodeRepository = new OppfolgingsPeriodeRepository(jdbcTemplate, transactor);

        authService = mock(AuthService.class);
        when(authService.getFnrOrThrow(AKTOR_ID)).thenReturn(FNR);
        bigQueryClient = mock(BigQueryClient.class);
        kafkaProducerService = mock(KafkaProducerService.class);
        manuellStatusService = mock(ManuellStatusService.class);

        oppfolgingService = new OppfolgingService(
                kafkaProducerService,
                null,
                null,
                authService,
                oppfolgingsStatusRepository,
                oppfolgingsPeriodeRepository,
                manuellStatusService,
                mock(AmtDeltakerClient.class),
                new KvpRepository(jdbcTemplate, namedParameterJdbcTemplate, transactor),
                new MaalRepository(jdbcTemplate, transactor),
                mock(BrukerOppslagFlereOppfolgingAktorRepository.class),
                transactor,
                mock(ArenaYtelserService.class),
                bigQueryClient,
                "https://test.nav.no"
        );

        startOppfolgingService = new StartOppfolgingService(
                manuellStatusService,
                oppfolgingsStatusRepository,
                oppfolgingsPeriodeRepository,
                kafkaProducerService,
                bigQueryClient,
                transactor,
                arenaOppfolgingService,
                "https://test.nav.no"
        );

        aktiverBrukerManueltService = new AktiverBrukerManueltService(
                authService, startOppfolgingService,
                DbTestUtils.createTransactor(jdbcTemplate)
        );


        DbTestUtils.cleanupTestDb();
        when(authService.getAktorIdOrThrow(any(Fnr.class))).thenReturn(AKTOR_ID);
        when(authService.getInnloggetVeilederIdent()).thenReturn("G321321");
        when(manuellStatusService.hentDigdirKontaktinfo(any())).thenReturn(new KRRData());
    }

    @Test
    public void skalLagreIDatabaseDersomKallTilArenaErOK() {
        startOppfolgingSomArbeidsoker(AKTOR_ID, FNR);
        Optional<Oppfolging> oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID);
        assertThat(oppfolging.isPresent()).isTrue();
    }

    @Test
    public void skalHaandtereAtOppfolgingstatusAlleredeFinnes() {
        oppfolgingsStatusRepository.opprettOppfolging(AKTOR_ID);
        oppfolgingsPeriodeRepository.avslutt(AKTOR_ID, "veilederid", "begrunnelse");
        startOppfolgingSomArbeidsoker(AKTOR_ID, FNR);
        Optional<Oppfolging> oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID);
        assertThat(oppfolging.get().isUnderOppfolging()).isTrue();
    }

    @Test
    public void aktiver_sykmeldt_skal_starte_oppfolging() {
        var oppfolgingFør = oppfolgingService.hentOppfolging(AKTOR_ID);
        assertThat(oppfolgingFør.isEmpty()).isTrue();
        aktiverBrukerManueltService.aktiverBrukerManuelt(FNR, "1234");
        var oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID);
        assertThat(oppfolging.get().isUnderOppfolging()).isTrue();
    }

}

File: src/test/kotlin/no/nav/veilarboppfolging/service/AktiverBrukerIntegrationTest.kt
package no.nav.veilarboppfolging.service

import no.nav.common.types.identer.AktorId
import no.nav.common.types.identer.EnhetId
import no.nav.common.types.identer.Fnr
import no.nav.common.types.identer.NavIdent
import no.nav.poao_tilgang.client.Decision
import no.nav.poao_tilgang.client.TilgangType
import no.nav.pto_schema.enums.arena.Formidlingsgruppe
import no.nav.veilarboppfolging.IntegrationTest
import no.nav.veilarboppfolging.oppfolgingsbruker.AvsluttetAvType
import no.nav.veilarboppfolging.oppfolgingsbruker.StartetAvType
import no.nav.veilarboppfolging.oppfolgingsbruker.inngang.OppfolgingStartBegrunnelse
import no.nav.veilarboppfolging.oppfolgingsperioderHendelser.hendelser.OppfolgingStartetHendelseDto
import no.nav.veilarboppfolging.oppfolgingsperioderHendelser.hendelser.OppfolgingsAvsluttetHendelseDto
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.within
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertInstanceOf
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.util.UUID

class AktiverBrukerIntegrationTest : IntegrationTest() {

    private val FNR: Fnr = Fnr.of("11111111111")
    private val AKTOR_ID: AktorId = AktorId.of("1234523423")

    @Test
    fun `skal lagre oppfolgingstatus på bruker når arbeidsoppfølging er startet`() {
        mockSytemBrukerAuthOk(AKTOR_ID, FNR)
        startOppfolgingSomArbeidsoker(AKTOR_ID, FNR)
        val oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID)
        Assertions.assertThat(oppfolging.isPresent()).isTrue()
    }

    @Test
    fun skalHaandtereAtOppfolgingstatusAlleredeFinnes() {
        mockSytemBrukerAuthOk(AKTOR_ID, FNR)
        oppfolgingsStatusRepository.opprettOppfolging(AKTOR_ID)
        oppfolgingsPeriodeRepository.avsluttSistePeriodeOgAvsluttOppfolging(AKTOR_ID, "veilederid", "begrunnelse")
        startOppfolgingSomArbeidsoker(AKTOR_ID, FNR)
        val oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID)
        Assertions.assertThat(oppfolging.get().isUnderOppfolging()).isTrue()
    }

    @Test
    fun aktiver_sykmeldt_skal_starte_oppfolging() {
        mockInternBrukerAuthOk(UUID.randomUUID(), AKTOR_ID, FNR)
        val oppfolgingFør = oppfolgingService.hentOppfolging(AKTOR_ID)
        Assertions.assertThat(oppfolgingFør.isEmpty()).isTrue()
        aktiverBrukerManueltService.aktiverBrukerManuelt(FNR, "1234")
        val oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID)
        Assertions.assertThat(oppfolging.get().isUnderOppfolging()).isTrue()
    }

    @Test
    fun `Skal lagre OppfolgingStartetHendelse-melding i utboks når oppfølging startes manuelt`() {
        val veilederIdent = NavIdent("B654321")
        val kontorSattAvVeileder = "4321"
        mockInternBrukerAuthOk(UUID.randomUUID(), AKTOR_ID, FNR, veilederIdent)
        val oppfolgingFør = oppfolgingService.hentOppfolging(AKTOR_ID)
        Assertions.assertThat(oppfolgingFør.isEmpty()).isTrue()

        aktiverBrukerManueltService.aktiverBrukerManuelt(FNR, kontorSattAvVeileder)

        val oppfolging = oppfolgingService.hentOppfolging(AKTOR_ID)
        val nyPeriode = oppfolging.get().oppfolgingsperioder.toList().maxByOrNull { it.startDato }!!
        val lagreteMeldingerIUtboks = getRecordsStoredInKafkaOutbox(kafkaProperties.oppfolgingshendelseV1, FNR.toString())
        assertThat(lagreteMeldingerIUtboks).hasSize(1)
        val hendelse = lagreteMeldingerIUtboks.first() as OppfolgingStartetHendelseDto
        assertThat(hendelse.fnr).isEqualTo(FNR.toString())
        assertThat(hendelse.startetBegrunnelse).isEqualTo(OppfolgingStartBegrunnelse.MANUELL_REGISTRERING_VEILEDER)
        assertThat(hendelse.arenaKontor).isNull()
        assertThat(hendelse.startetAvType).isEqualTo(StartetAvType.VEILEDER)
        assertThat(hendelse.startetAv).isEqualTo(veilederIdent)
        assertThat(hendelse.startetTidspunkt).isCloseTo(ZonedDateTime.now(), within(1, ChronoUnit.SECONDS))
        assertThat(hendelse.foretrukketArbeidsoppfolgingskontor).isEqualTo(kontorSattAvVeileder)
        assertThat(hendelse.oppfolgingsPeriodeId).isEqualTo(nyPeriode.uuid)
    }

    @Test
    fun `Skal lagre OppfolgingStartetHendelse-melding i utboks når oppfølging av arbeidssøkerregistrering startes`() {
        val veilederIdent = NavIdent("B654321")
        mockInternBrukerAuthOk(UUID.randomUUID(), AKTOR_ID, FNR, veilederIdent)
        val oppfolgingFør = oppfolgingService.hentOppfolging(AKTOR_ID)
        assertThat(oppfolgingFør.isEmpty).isTrue()

        startOppfolgingSomArbeidsoker(AKTOR_ID, FNR)

        val lagreteMeldingerIUtboks = getRecordsStoredInKafkaOutbox(kafkaProperties.oppfolgingshendelseV1, FNR.toString())
        assertThat(lagreteMeldingerIUtboks).hasSize(1)
        val hendelse = lagreteMeldingerIUtboks.first() as OppfolgingStartetHendelseDto
        assertThat(hendelse.startetBegrunnelse).isEqualTo(OppfolgingStartBegrunnelse.ARBEIDSSOKER_REGISTRERING)
        assertThat(hendelse.arenaKontor).isNull()
        assertThat(hendelse.foretrukketArbeidsoppfolgingskontor).isNull()
    }

    @Test
    fun `Skal lagre OppfolgingAvsluttetHendelse-melding i utboks når oppfølging avsluttes`() {
        val veilederIdent = NavIdent("B654321")
        val veilederUUID = UUID.randomUUID()
        val enhetId = EnhetId.of("3333")
        val avsluttBegrunnelse = "avslutt begrunnelse"
        mockInternBrukerAuthOk(veilederUUID, AKTOR_ID, FNR, veilederIdent)
        mockPoaoTilgangHarTilgangTilBruker(veilederUUID, FNR, Decision.Permit, TilgangType.SKRIVE)
        mockPoaoTilgangHarTilgangTilEnhet(veilederUUID, enhetId)
        mockVeilarbArenaOppfolgingsBruker(FNR, Formidlingsgruppe.ISERV, oppfolgingsEnhet = enhetId.get())

        startOppfolgingSomArbeidsoker(AKTOR_ID, FNR)
        avsluttOppfolgingManueltSomVeileder(AKTOR_ID, veilederIdent, begrunnelse = avsluttBegrunnelse)

        val lagreteMeldingerIUtboks = getRecordsStoredInKafkaOutbox(kafkaProperties.oppfolgingshendelseV1, FNR.toString())
        assertThat(lagreteMeldingerIUtboks).hasSize(2)
        val startHendelse = lagreteMeldingerIUtboks.first()
        val avsluttetHendelse = lagreteMeldingerIUtboks.last()
        assertInstanceOf<OppfolgingStartetHendelseDto>(startHendelse)
        assertInstanceOf<OppfolgingsAvsluttetHendelseDto>(avsluttetHendelse)
        assertThat(avsluttetHendelse.fnr).isEqualTo(FNR.toString())
        assertThat(avsluttetHendelse.avsluttetAv).isEqualTo(veilederIdent)
        assertThat(avsluttetHendelse.avsluttetAvType).isEqualTo(AvsluttetAvType.VEILEDER)
        assertThat(avsluttetHendelse.startetTidspunkt).isEqualTo(startHendelse.startetTidspunkt)
        assertThat(avsluttetHendelse.avsluttetTidspunkt).isCloseTo(ZonedDateTime.now(), within(1, ChronoUnit.SECONDS))
        assertThat(avsluttetHendelse.oppfolgingsPeriodeId).isEqualTo(startHendelse.oppfolgingsPeriodeId)
    }
}
